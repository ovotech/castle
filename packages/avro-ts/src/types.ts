import * as ts from 'typescript';
import { Document, DocumentContext } from '@ovotech/ts-compose';
import { Schema } from 'avsc';

export type CustomLogicalType =
  | { module: string; named: string }
  | { module: string; defaultAs: string }
  | { module: string; allAs: string };

export interface Context extends DocumentContext {
  logicalTypes?: { [key: string]: string | CustomLogicalType };
  namespace?: string;
  refs?: { [key: string]: Schema };
  external?: { [file: string]: { [key: string]: Schema } };
  defaultsAsOptional?: boolean;
}

export type Convert<TSchema = Schema, TType = ts.TypeNode> = (
  context: Context,
  type: TSchema,
) => Document<TType, Context>;
