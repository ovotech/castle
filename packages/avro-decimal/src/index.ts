import { Type, Schema, types } from 'avsc';
import { Decimal } from 'decimal.js';
import { Int64BE } from 'int64-buffer';

export function decimalToBuffer(val: Decimal, scale: number): Buffer {
  const scaled = val.mul(10 ** scale);
  const value = new Int64BE(scaled.trunc().toString());
  return value.toBuffer();
}

export function to64Bit(buf: Buffer): Buffer {
  const isNegatve = buf[0] & 0x80;
  const toFill = 8 - buf.length;

  const bufferElements = Array.prototype.slice.call(buf, 0);

  return Buffer.from([
    ...[...new Array(toFill)].map(() => (isNegatve ? 0xff : 0x00)),
    ...bufferElements,
  ]);
}

export function bufferToDecimal(buf: Buffer, scale: number): Decimal {
  const unscaled = new Int64BE(to64Bit(buf));
  return new Decimal(unscaled.toString()).div(10 ** scale);
}

export interface SchemaOptions {
  precision: number;
  scale: number;
}

export class AvroDecimal extends types.LogicalType {
  public precision: number;
  public scale: number;

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types,@typescript-eslint/no-explicit-any
  public constructor(schema: any, opts?: unknown) {
    super(schema, opts);
    this.precision = schema.precision;
    this.scale = schema.scale;
  }

  public _export(attrs: Schema & SchemaOptions): void {
    attrs.precision = this.precision;
    attrs.scale = this.scale;
  }

  public _resolve(type: Type & SchemaOptions): (<T>(x: T) => T) | undefined {
    if (
      type instanceof AvroDecimal &&
      Type.isType(type, 'logical:decimal') &&
      type.precision === this.precision &&
      type.scale === this.scale
    ) {
      return <T>(x: T): T => x;
    } else {
      return undefined;
    }
  }

  public _fromValue(buf: unknown): Decimal {
    if (!(buf instanceof Buffer)) {
      throw new Error('expecting underlying Buffer type');
    }

    if (buf.length > 8) {
      throw new Error('buffers with more than 64bits are not supported');
    }

    return bufferToDecimal(buf, this.scale);
  }

  public _toValue(val: unknown): Buffer {
    // Workaround for `val instanceof Decimal` returning false negavites
    // across versions of ts, gives some level of sanity checking/better
    // error when accidentally passing in garbage
    const dVal = val as Decimal;
    if (!dVal.mul) {
      throw new Error('expecting Decimal type');
    }

    return decimalToBuffer(dVal, this.scale);
  }
}
