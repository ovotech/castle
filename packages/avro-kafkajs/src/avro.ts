import { SchemaRegistry, DecodeItemParams } from './SchemaRegistry';
import { Schema } from 'avsc';
import {
  AvroMessage,
  AvroProducerRecord,
  AvroProducerBatch,
  AvroEachMessage,
  AvroEachBatch,
  AvroKafkaMessage,
  AvroEachBatchPayload,
  TopicsAlias,
} from './types';
import {
  Message,
  ProducerRecord,
  ProducerBatch,
  TopicMessages,
  EachMessagePayload,
  EachBatchPayload,
  KafkaMessage,
} from 'kafkajs';

export const encodeMessages = async <T = unknown, KT = unknown>({
  schemaRegistry,
  avroMessages,
  ...params
}: {
  schemaRegistry: SchemaRegistry;
  avroMessages: AvroMessage<T, KT>[];
} & (
  | { schema: Schema; keySchema?: Schema; topic: string }
  | { subject: string; keySubject?: string }
)): Promise<Message[]> => {
  const messages: Message[] = [];

  const keyParams: DecodeItemParams | undefined =
    'subject' in params
      ? params.keySubject
        ? { subject: params.keySubject }
        : undefined
      : params.keySchema
      ? { topic: params.topic, schemaType: 'key', schema: params.keySchema }
      : undefined;

  const valueParams: DecodeItemParams =
    'subject' in params
      ? { subject: params.subject }
      : { topic: params.topic, schemaType: 'value', schema: params.schema };

  for (const message of avroMessages) {
    messages.push({
      ...message,
      key:
        keyParams && message.key
          ? await schemaRegistry.encode<KT>({ ...keyParams, value: message.key })
          : // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (message.key as any),
      value: await schemaRegistry.encode<T>({ ...valueParams, value: message.value }),
    });
  }
  return messages;
};

export const toProducerRecord = async <T = unknown, KT = Message['key']>(
  schemaRegistry: SchemaRegistry,
  record: AvroProducerRecord<T, KT>,
): Promise<ProducerRecord> => {
  const { messages: avroMessages, ...rest } = record;
  const messages = await encodeMessages<T, KT>({ schemaRegistry, avroMessages, ...rest });

  return { ...rest, messages };
};

export const toProducerBatch = async (
  schemaRegistry: SchemaRegistry,
  record: AvroProducerBatch,
): Promise<ProducerBatch> => {
  const topicMessages: TopicMessages[] = [];
  const { topicMessages: avroTopicMessages, ...rest } = record;
  for (const item of avroTopicMessages) {
    const { messages: avroMessages, ...itemRest } = item;
    const messages = await encodeMessages({ schemaRegistry, avroMessages, ...itemRest });
    topicMessages.push({ ...itemRest, messages });
  }
  return { ...rest, topicMessages };
};

export const toAvroEachMessage = <T = unknown, KT = KafkaMessage['key']>(
  schemaRegistry: SchemaRegistry,
  eachMessage: AvroEachMessage<T, KT>,
  encodedKey?: boolean,
) => {
  return async (payload: EachMessagePayload): Promise<void> => {
    const key =
      encodedKey && payload.message.key
        ? await schemaRegistry.decode<KT>(payload.message.key)
        : payload.message.key;

    if (payload.message.value !== null) {
      const { type, value } = await schemaRegistry.decodeWithType<T>(payload.message.value);

      return eachMessage({
        ...payload,
        message: { ...payload.message, key: key as KT, value, schema: type.schema() },
      });
    }
  };
};

export const toAvroEachBatch = <T = unknown, KT = KafkaMessage['key']>(
  schemaRegistry: SchemaRegistry,
  eachBatch: AvroEachBatch<T, KT>,
  encodedKey?: boolean,
) => {
  return async (payload: EachBatchPayload): Promise<void> => {
    const avroPayload = (payload as unknown) as AvroEachBatchPayload<T, KT>;
    const messages: AvroKafkaMessage<T, KT>[] = [];
    for (const message of payload.batch.messages) {
      if (message.value) {
        const { value, type } = await schemaRegistry.decodeWithType<T>(message.value);
        const key =
          encodedKey && message.key ? await schemaRegistry.decode<KT>(message.key) : message.key;

        messages.push({ ...message, key: key as KT, value, schema: type.schema() });
      }
    }
    avroPayload.batch.messages = messages;
    return eachBatch(avroPayload);
  };
};

export const resolveTopic = <T extends { topic: unknown }>(
  record: T,
  topicsAlias: TopicsAlias,
): T => ({
  ...record,
  topic:
    typeof record.topic === 'string' && topicsAlias[record.topic]
      ? topicsAlias[record.topic]
      : record.topic,
});
