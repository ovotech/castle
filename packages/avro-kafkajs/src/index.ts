export {
  SchemaRegistry,
  AvroBuffer,
  DecodeItem,
  EncodeCache,
  DecodeCache,
  SchemaRegistryConfig,
  SchemaVersion,
} from './SchemaRegistry';
export { AvroKafka } from './AvroKafka';
export { AvroProducer } from './AvroProducer';
export { AvroConsumer } from './AvroConsumer';
export { AvroTransaction } from './AvroTransaction';
export {
  AvroTransformBatch,
  AvroTransformBatchConfig,
  ToPartition,
  ToKafkaMessage,
  StreamKafkaMessage,
} from './AvroTransformBatch';
export {
  encodeMessages,
  toProducerRecord,
  toProducerBatch,
  toAvroEachMessage,
  toAvroEachBatch,
  resolveTopic,
} from './avro';
export {
  AvroEachMessagePayload,
  AvroBatch,
  AvroEachBatchPayload,
  AvroMessage,
  AvroKafkaMessage,
  AvroProducerRecord,
  AvroTopicMessages,
  AvroProducerBatch,
  AvroProducerRecordSchema,
  AvroConsumerRun,
  TopicsAlias,
} from './types';
