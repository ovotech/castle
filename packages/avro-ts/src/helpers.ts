import { Context } from './types';
import { schema as avroSchema } from 'avsc';
import { Document, Node, withIdentifier, document, Type } from '@ovotech/ts-compose';
import * as ts from 'typescript';

export const fullName = (
  context: Context,
  schema: avroSchema.RecordType | avroSchema.EnumType,
): string => `${schema.namespace ?? context.namespace}.${schema.name}`;

export const firstUpperCase = (name: string): string =>
  name ? name[0].toUpperCase() + name.slice(1) : name;

export const convertNamespace = (namespace: string): string =>
  namespace.split('.').map(firstUpperCase).join('');

export const nameParts = (fullName: string): [string] | [string, string] => {
  const parts = fullName.split('.');
  return parts.length > 1
    ? [parts.slice(parts.length - 1)[0], parts.slice(0, parts.length - 1).join('.')]
    : [parts[0]];
};

export const namedType = (
  type: ts.InterfaceDeclaration | ts.TypeAliasDeclaration,
  context: Context,
  schema: avroSchema.RecordType | avroSchema.EnumType,
  namespace?: string,
): Document<ts.TypeNode, Context> => {
  const name = firstUpperCase(schema.name);
  const namespaceName = namespace ? convertNamespace(namespace) : undefined;
  const fullName = namespaceName ? [namespaceName, name] : name;

  const contextWithRef = namespace
    ? withIdentifier(
        context,
        Node.Const({
          name: `${name}Name`,
          isExport: true,
          value: `${namespace}.${schema.name}`,
        }),
        namespaceName,
      )
    : context;

  return document(withIdentifier(contextWithRef, type, namespaceName), Type.Referance(fullName));
};
