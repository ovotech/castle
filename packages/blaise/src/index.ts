import { AvroKafkaMessage } from '@ovotech/avro-kafkajs/dist/types';
import avroMock, { Seeded } from '@ovotech/avro-mock-generator';
import {
  CastleEachBatchPayload,
  CastleEachMessagePayload,
} from '@ovotech/castle';
import merge = require('lodash.merge');
import { DeepPartial } from 'ts-essentials';
import { BlaiseDefaults, Blaise } from './types';

// eslint-disable-next-line @typescript-eslint/no-empty-function, @typescript-eslint/explicit-function-return-type
const noop = () => {};

/* eslint-disable @typescript-eslint/no-explicit-any */
export const defaultPayload: BlaiseDefaults = {
  eachMessage: {
    topic: 'blaise-default-topic',
    partition: 0,
    producer: noop as any,
  },
  eachBatch: {
    batch: {
      topic: '',
      highWatermark: '',
      partition: 0,
      isEmpty: noop as any,
      firstOffset: noop as any,
      lastOffset: noop as any,
      offsetLag: noop as any,
      offsetLagLow: noop as any,
    },
    isStale: noop as any,
    isRunning: noop as any,
    heartbeat: noop as any,
    uncommittedOffsets: noop as any,
    producer: noop as any,
    resolveOffset: noop as any,
    commitOffsetsIfNecessary: noop as any,
  },
  message: {
    schema: [],
    key: Buffer.from('blaise-default-key'),
    value: {},
    timestamp: Date.now().toString(),
    size: 3,
    attributes: 3,
    offset: '1',
  },
  avro: {
    schema: 'boolean',
  },
  generator: avroMock,
};
/* eslint-enable @typescript-eslint/no-explicit-any */

const buildBlaise = <T = unknown>(defaults = defaultPayload): Blaise<T> => {
  const blaise = <T1 = T>(
    newDefaults?: DeepPartial<BlaiseDefaults>,
  ): Blaise<T1> => {
    if (newDefaults?.avro?.seed) {
      newDefaults.generator = Seeded<T>(newDefaults.avro.seed);
    }
    return buildBlaise<T1>(merge({}, defaults, newDefaults));
  };

  blaise.default = blaise;

  blaise.pickUnion = <T1 = T>(pickUnion: Array<string>) =>
    blaise<T1>({ avro: { pickUnion } });

  blaise.seed = <T1 = T>(seed: number) => blaise<T1>({ avro: { seed } });

  blaise.getDefault = () => defaults as BlaiseDefaults<T>;

  blaise.message = (
    message?: DeepPartial<AvroKafkaMessage<T>>,
  ): AvroKafkaMessage<T> =>
    merge(
      {
        value: defaults.generator(defaults.avro.schema, defaults.avro),
      },
      defaults.message,
      message,
    ) as AvroKafkaMessage<T>;

  blaise.eachMessage = (
    message?: DeepPartial<AvroKafkaMessage<T>>,
  ): CastleEachMessagePayload<T> => ({
    ...defaults.eachMessage,
    message: blaise.message(message),
  });

  blaise.eachBatch = (
    messages: Array<AvroKafkaMessage<T>> = [],
  ): CastleEachBatchPayload<T> =>
    merge({}, defaults.eachBatch, {
      batch: { messages },
    });

  return blaise;
};

export default buildBlaise();
