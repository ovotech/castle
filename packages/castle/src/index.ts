export {
  createCastle,
  produce,
  consumeEachMessage,
  consumeEachBatch,
  optionalConsumers,
  createKafka,
  createProducer,
  createConsumers,
  createCastleFromParts,
} from './castle';
export {
  CastleEachMessagePayload,
  CastleEachBatchPayload,
  Resolver,
  Middleware,
  CastleConsumerConfig,
  CastleConsumer,
  CastleConfig,
  Castle,
} from './types';
export { describeCastle } from './describe';
export {
  createLogging,
  toLogCreator,
  LoggingContext,
  LoggerOptions,
  Logger,
} from './middlewares/logging';
