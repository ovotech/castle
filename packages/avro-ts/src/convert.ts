import { Schema, schema } from 'avsc';
import { document, Type, printDocument } from '@ovotech/ts-compose';
import { Convert, Context } from './types';
import { isWrappedUnion, convertWrappedUnionType } from './types/wrapped-union';
import { isUnion, convertUnionType } from './types/union';
import { isLogicalType, convertLogicalType } from './types/logical';
import { isRecordType, convertRecordType } from './types/record';
import { isArrayType, convertArrayType } from './types/array';
import { isMapType, convertMapType } from './types/map';
import { isEnumType, convertEnumType } from './types/enum';
import { isPrimitiveType, convertPrimitiveType } from './types/primitive';
import { isFixedType, convertFixedType } from './types/fixed';
import { withHeader, withImports } from '@ovotech/ts-compose/dist/document';
import { fullName, firstUpperCase, nameParts, convertNamespace } from './helpers';
import * as ts from 'typescript';

export const addRef = (type: schema.RecordType | schema.EnumType, context: Context): Context => ({
  ...context,
  namespace: type.namespace ?? context.namespace,
  refs: {
    ...context.refs,
    [fullName(context, type)]: { ...type, namespace: type.namespace ?? context.namespace },
  },
});

export const collectRefs = (type: Schema, context: Context): Context => {
  if (isUnion(type)) {
    return type.reduce((all, item) => collectRefs(item, all), context);
  } else if (isArrayType(type)) {
    return collectRefs(type.items, context);
  } else if (isMapType(type)) {
    return collectRefs(type.values, context);
  } else if (isEnumType(type)) {
    return addRef(type, context);
  } else if (isRecordType(type)) {
    return type.fields.reduce((all, item) => collectRefs(item.type, all), addRef(type, context));
  } else {
    return context;
  }
};

export const convertType: Convert = (context, type) => {
  if (isWrappedUnion(type, context)) {
    return convertWrappedUnionType(context, type);
  } else if (isUnion(type)) {
    return convertUnionType(context, type);
  } else if (isLogicalType(type)) {
    return convertLogicalType(context, type);
  } else if (isRecordType(type)) {
    return convertRecordType(context, type);
  } else if (isArrayType(type)) {
    return convertArrayType(context, type);
  } else if (isMapType(type)) {
    return convertMapType(context, type);
  } else if (isEnumType(type)) {
    return convertEnumType(context, type);
  } else if (isFixedType(type)) {
    return convertFixedType(context, type);
  } else if (isPrimitiveType(type)) {
    return convertPrimitiveType(context, type);
  } else if (typeof type === 'string') {
    const [name, nameNamespace] = nameParts(type);
    const namespace = nameNamespace ?? context.namespace;

    if (namespace && context.external && !context.refs?.[type]) {
      for (const module in context.external) {
        if (context.external[module][type]) {
          const externalNamespace = convertNamespace(namespace);
          const alias = `${externalNamespace}${firstUpperCase(name)}`;
          const externalContext = withImports(context, {
            named: [{ name: convertNamespace(namespace), as: alias }],
            module,
          });
          const ref = Type.Referance([alias, firstUpperCase(name)]);
          return document(externalContext, ref);
        }
      }
    }

    const ref = namespace
      ? Type.Referance([convertNamespace(namespace), firstUpperCase(name)])
      : Type.Referance(firstUpperCase(name));

    return document(context, ref);
  } else {
    throw new Error(`Cannot work out type ${type}`);
  }
};

export const toTypeScript = (schema: Schema, initial: Context = {}): string => {
  const contextWithRefs = collectRefs(schema, initial);
  const { context, type } = convertType(contextWithRefs, schema);

  const contextWithHeader = context.namespaces
    ? withHeader(context, '/* eslint-disable @typescript-eslint/no-namespace */')
    : context;

  const name =
    ts.isTypeReferenceNode(type) && ts.isQualifiedName(type.typeName)
      ? type.typeName.right
      : 'AvroType';

  return printDocument(document(contextWithHeader, Type.Alias({ name, isExport: true, type })));
};

export const toExternalContext = (
  schema: Schema,
  initial: Context = {},
): { [key: string]: Schema } => {
  const contextWithRefs = collectRefs(schema, initial);
  const { context } = convertType(contextWithRefs, schema);
  return context.refs ?? {};
};
