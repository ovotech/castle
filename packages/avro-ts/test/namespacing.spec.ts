import { schema, Type } from 'avsc';
import { toTypeScript } from '../src';

describe('Avro simple and full name related tests', () => {
  it.each<[string, { schema: schema.RecordType; testObject: unknown }]>([
    [
      'there are no namespaces',
      {
        schema: {
          type: 'record',
          name: 'RootRecord',
          fields: [
            {
              name: 'unionField',
              type: [
                {
                  type: 'record',
                  name: 'SomeRecordA',
                  fields: [
                    {
                      name: 'someRecordAField',
                      type: 'string',
                    },
                  ],
                },
                {
                  type: 'record',
                  name: 'SomeRecordB',
                  fields: [
                    {
                      name: 'someRecordBField',
                      type: 'string',
                    },
                  ],
                },
              ],
            },
          ],
        } as schema.RecordType,
        testObject: {
          unionField: {
            SomeRecordA: {
              someRecordAField: 'test',
            },
          },
        },
      },
    ],
    [
      'inheriting a namespace',
      {
        schema: {
          type: 'record',
          name: 'RootRecord',
          namespace: 'RootNamespace',
          fields: [
            {
              name: 'unionField',
              type: [
                {
                  type: 'record',
                  name: 'SomeRecordA',
                  fields: [
                    {
                      name: 'someRecordAField',
                      type: 'string',
                    },
                  ],
                },
                {
                  type: 'record',
                  name: 'SomeRecordB',
                  fields: [
                    {
                      name: 'someRecordBField',
                      type: 'string',
                    },
                  ],
                },
              ],
            },
          ],
        },
        testObject: {
          unionField: {
            ['RootNamespace.SomeRecordA']: {
              someRecordAField: 'test',
            },
          },
        },
      },
    ],
    [
      'specifying a full name',
      {
        schema: {
          type: 'record',
          name: 'RootRecord',
          fields: [
            {
              name: 'unionField',
              type: [
                {
                  type: 'record',
                  name: 'RecordANamespace.SomeRecordA',
                  fields: [
                    {
                      name: 'someRecordAField',
                      type: 'string',
                    },
                  ],
                },
                {
                  type: 'record',
                  name: 'RecordBNamespace.SomeRecordB',
                  fields: [
                    {
                      name: 'someRecordBField',
                      type: 'string',
                    },
                  ],
                },
              ],
            },
          ],
        },
        testObject: {
          unionField: {
            ['RecordANamespace.SomeRecordA']: {
              someRecordAField: 'test',
            },
          },
        },
      },
    ],
  ])('Should properly generate union type names when %s', (description, { schema, testObject }) => {
    const tsCode = toTypeScript(schema);

    const type = Type.forSchema(schema);
    const result = type.fromBuffer(type.toBuffer(testObject));

    // Adding the actual decoding result to visually make sure the types actually match the
    // avro decoder behavior.
    expect(result).toMatchSnapshot();
    expect(tsCode).toMatchSnapshot();
  });
});
