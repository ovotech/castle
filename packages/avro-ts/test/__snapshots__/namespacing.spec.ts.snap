// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avro simple and full name related tests Should properly generate union type names when inheriting a namespace 1`] = `
RootRecord {
  "unionField": Branch$ {
    "RootNamespace.SomeRecordA": SomeRecordA {
      "someRecordAField": "test",
    },
  },
}
`;

exports[`Avro simple and full name related tests Should properly generate union type names when inheriting a namespace 2`] = `
"/* eslint-disable @typescript-eslint/no-namespace */

export type RootRecord = RootNamespace.RootRecord;

export namespace RootNamespace {
    export const SomeRecordASchema = \\"{\\\\\\"type\\\\\\":\\\\\\"record\\\\\\",\\\\\\"name\\\\\\":\\\\\\"SomeRecordA\\\\\\",\\\\\\"fields\\\\\\":[{\\\\\\"name\\\\\\":\\\\\\"someRecordAField\\\\\\",\\\\\\"type\\\\\\":\\\\\\"string\\\\\\"}]}\\";
    export const SomeRecordAName = \\"RootNamespace.SomeRecordA\\";
    export interface SomeRecordA {
        someRecordAField: string;
    }
    export const SomeRecordBSchema = \\"{\\\\\\"type\\\\\\":\\\\\\"record\\\\\\",\\\\\\"name\\\\\\":\\\\\\"SomeRecordB\\\\\\",\\\\\\"fields\\\\\\":[{\\\\\\"name\\\\\\":\\\\\\"someRecordBField\\\\\\",\\\\\\"type\\\\\\":\\\\\\"string\\\\\\"}]}\\";
    export const SomeRecordBName = \\"RootNamespace.SomeRecordB\\";
    export interface SomeRecordB {
        someRecordBField: string;
    }
    export const RootRecordSchema = \\"{\\\\\\"type\\\\\\":\\\\\\"record\\\\\\",\\\\\\"name\\\\\\":\\\\\\"RootRecord\\\\\\",\\\\\\"namespace\\\\\\":\\\\\\"RootNamespace\\\\\\",\\\\\\"fields\\\\\\":[{\\\\\\"name\\\\\\":\\\\\\"unionField\\\\\\",\\\\\\"type\\\\\\":[{\\\\\\"type\\\\\\":\\\\\\"record\\\\\\",\\\\\\"name\\\\\\":\\\\\\"SomeRecordA\\\\\\",\\\\\\"fields\\\\\\":[{\\\\\\"name\\\\\\":\\\\\\"someRecordAField\\\\\\",\\\\\\"type\\\\\\":\\\\\\"string\\\\\\"}]},{\\\\\\"type\\\\\\":\\\\\\"record\\\\\\",\\\\\\"name\\\\\\":\\\\\\"SomeRecordB\\\\\\",\\\\\\"fields\\\\\\":[{\\\\\\"name\\\\\\":\\\\\\"someRecordBField\\\\\\",\\\\\\"type\\\\\\":\\\\\\"string\\\\\\"}]}]}]}\\";
    export const RootRecordName = \\"RootNamespace.RootRecord\\";
    export interface RootRecord {
        unionField: {
            \\"RootNamespace.SomeRecordA\\": RootNamespace.SomeRecordA;
            \\"RootNamespace.SomeRecordB\\"?: never;
        } | {
            \\"RootNamespace.SomeRecordA\\"?: never;
            \\"RootNamespace.SomeRecordB\\": RootNamespace.SomeRecordB;
        };
    }
}
"
`;

exports[`Avro simple and full name related tests Should properly generate union type names when specifying a full name 1`] = `
RootRecord {
  "unionField": Branch$ {
    "RecordANamespace.SomeRecordA": SomeRecordA {
      "someRecordAField": "test",
    },
  },
}
`;

exports[`Avro simple and full name related tests Should properly generate union type names when specifying a full name 2`] = `
"export type AvroType = RootRecord;

export interface RecordANamespaceSomeRecordA {
    someRecordAField: string;
}

export interface RecordBNamespaceSomeRecordB {
    someRecordBField: string;
}

export interface RootRecord {
    unionField: {
        \\"RecordANamespace.SomeRecordA\\": RecordANamespaceSomeRecordA;
        \\"RecordBNamespace.SomeRecordB\\"?: never;
    } | {
        \\"RecordANamespace.SomeRecordA\\"?: never;
        \\"RecordBNamespace.SomeRecordB\\": RecordBNamespaceSomeRecordB;
    };
}
"
`;

exports[`Avro simple and full name related tests Should properly generate union type names when there are no namespaces 1`] = `
RootRecord {
  "unionField": Branch$ {
    "SomeRecordA": SomeRecordA {
      "someRecordAField": "test",
    },
  },
}
`;

exports[`Avro simple and full name related tests Should properly generate union type names when there are no namespaces 2`] = `
"export type AvroType = RootRecord;

export interface SomeRecordA {
    someRecordAField: string;
}

export interface SomeRecordB {
    someRecordBField: string;
}

export interface RootRecord {
    unionField: {
        SomeRecordA: SomeRecordA;
        SomeRecordB?: never;
    } | {
        SomeRecordA?: never;
        SomeRecordB: SomeRecordB;
    };
}
"
`;
